### ISSUE TITLE
undesired side effect: can it be removed?

### ISSUE THREAD

**Issue by pjljvandelaar** [CVC4/CVC4#2157]
We use problem solvers to handle constraints our users add in their test models.
On top of that we add randomization since in tests we want to vary the input as much as possible.

We however noticed that some numbers are hardly chosen. Why? 
Before trying to randomize we ask the problem solver: does a solution exists with another value?
Since if another solution does not exists, trying to find a random solution is a waste of time.

Based on our experience I generated the follow smt2 input that show the issue:
```
(get-info :name)
(get-info :version)
(set-option :produce-models true)
(set-logic ALL)
(set-info :smt-lib-version 2.5)

(define-fun
    f1044$validInt ( (x$$1045 Int) ) Bool
    (and (<= 0 x$$1045) (<= 0 (+ 10 (- x$$1045))))
)
(push 1)
(declare-fun $In$1050$0$1$() Int)
(assert (f1044$validInt $In$1050$0$1$))
(check-sat)
(get-value ($In$1050$0$1$))

(push 1)
(assert (and (not (<= 0 (+ (- 2) (- $In$1050$0$1$)))) (<= 0 (+ 4 (- $In$1050$0$1$)))))
(check-sat)
(get-value ($In$1050$0$1$))
(pop 1)

(push 1)
(assert (not (= 0 $In$1050$0$1$)))
(check-sat)
(pop 1)

(push 1)
(assert (and (not (<= 0 (+ (- 2) (- $In$1050$0$1$)))) (<= 0 (+ 4 (- $In$1050$0$1$)))))
(check-sat)
(get-value ($In$1050$0$1$))
(pop 1)

(push 1)
(assert (and (not (<= 0 (+ (- 2) (- $In$1050$0$1$)))) (<= $In$1050$0$1$ 0)))
(check-sat)
(get-value ($In$1050$0$1$))
(pop 1)
(pop 1)
```

And we get the answers:
```
C:\experiments> cvc4 --incremental .\distributionPushPop.smt2
(:name "cvc4")
(:version "1.6")
sat
(($In$1050$0$1$ 0))
sat
(($In$1050$0$1$ 0))
sat
sat
(($In$1050$0$1$ 1))
sat
(($In$1050$0$1$ 0))
```

In other words:
```
(push 1)
(assert (not (= 0 $In$1050$0$1$)))
(check-sat)
(pop 1)
```
has a side effect on the later statement (yet since it is popped I had not expected this!)

Although the answer is correct, and no bug is present (see the last test that 0 is indeed also valid),
I wonder whether it is possible to remove this undesired side effect?

For your information: this behavior is also observed in z3. However, Yices does exhibit the desired behavior so it seems possible.

**Comment by 4tXJ7f** [comment]
It actually looks like in the case of CVC4, this behavior is coming from the arithmetic engine. I translated the problem to an equivalent BV problem and got 0 as a result every time. I could also imagine that some solvers, such as Z3, just try to plug in the last model and see if that works for the new constraints (assuming that there haven't been any new variables) but that is not the case for CVC4. I can do some further investigation into why/where the arithmetic engine is doing this but I'd like to understand your motivation a bit better. The issue is that even if we change the behavior in this version of CVC4, we might introduce a similar issue in the future (or even answer 1 after the first check-sat call) because CVC4 only concerns itself with finding *a* model not the model with the smallest values. From the commit that refers to this issue, it seems like you are trying to get some upper bound that is as tight as possible? Have you considered using quantifiers to ensure that you have the tightest bound? Alternatively, you could issue a second query (or multiple queries) asserting that the value must be smaller than the last solution.

**Comment by pjljvandelaar** [comment]
@4tXJ7f you are right CVC4 (like Z3) only cares about getting a solution. My additional requirement to treat all possible solutions equally is an additional requirement that can't be easily combined.
In the above example, in the statements that are popped, CVC4 might have also learned something on the earlier defined functions: so side effects are even desired. Keeping this knowledge can speed up an answer but also biases solving the next problems. So I will close the issue, since it is not relevant/essential for the problem solver. Yet, I as a specific user with a specific use case has to solve it in my context.

**Comment by martin-cs** [comment]
On Sun, 2018-07-08 at 23:40 -0700, pjljvandelaar wrote:
> @4tXJ7f you are right CVC4 (like Z3) only cares about getting a
> solution. My additional requirement to treat all possible solutions
> equally is an additional requirement that can't be easily combined.
> In the above example, in the statements that are popped, CVC4 might
> have also learned something on the earlier defined functions: so side
> effects are even desired. Keeping this knowledge can speed up an
> answer but also biases solving the next problems.

This is one of many.  Things like the phase saving of the SAT solver
will add biases for significant performance improvements.


>  So I will close the issue, since it is not relevant/essential for
> the problem solver. Yet, I as a specific user with a specific use
> case has to solve it in my context.

Some of the work on #SAT and #SMT have used uniform hashing to produce
reasonably robust sampling of the space of models.  This may be the
kind of thing you want but, to the best of my understanding that is
best implemented outside of the solver (but may be generic and able to
be shared between users / applications).
